src/python/pants/backend/jvm/subsystems/graal.py

```diff
@@ -0,0 +1,182 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+import os
+import shutil
+
+from pants.backend.native.config.environment import Platform
+from pants.base.build_environment import get_pants_cachedir
+from pants.base.hash_utils import hash_file, stable_json_sha1
+from pants.base.workunit import WorkUnitLabel
+from pants.binaries.binary_tool import NativeTool
+from pants.binaries.binary_util import BinaryToolUrlGenerator
+from pants.option.compiler_option_sets_mixin import CompilerOptionSetsMixin
+from pants.util.contextutil import temporary_dir
+from pants.util.dirutil import safe_mkdir_for
+from pants.util.memo import memoized_method, memoized_property
+from pants.util.meta import classproperty
+from pants.util.process_handler import subprocess
+from pants.util.strutil import safe_shlex_join
+
+
+logger = logging.getLogger(__name__)
+
+
+class GraalCEUrlGenerator(BinaryToolUrlGenerator):
+
+  _DIST_URL_FMT = 'https://github.com/oracle/graal/releases/download/vm-{version}/{base}'
+  _ARCHIVE_BASE_FMT = 'graalvm-ce-{version}-{system_id}-amd64.tar.gz'
+  _SYSTEM_ID = {
+    'mac': 'macos',
+    'linux': 'linux',
+  }
+
+  def generate_urls(self, version, host_platform):
+    system_id = self._SYSTEM_ID[host_platform.os_name]
+    archive_basename = self._ARCHIVE_BASE_FMT.format(version=version, system_id=system_id)
+    return [self._DIST_URL_FMT.format(version=version, base=archive_basename)]
```

Comment:
This uses the same resolution mechanism as other binary tools, in that if `--no-allow-external-binary-tool-downloads` is set, it will attempt to fetch it from `--binaries-baseurls`. The url generators are a convenience for open source users which allows someone to set e.g. `--graal-version=1.0.0rc16` when that version is released in the future, without requiring any work on our part to allow that. We haven't yet upstreamed any script which syncs our pantsbuild s3 to another source, so people in other monorepos without public internet access will have to write that code for themselves.